Src/Common/CorrelationIdLookupHelper.cs

```diff
@@ -119,29 +121,49 @@ public bool TryGetXComponentCorrelationId(string instrumentationKey, out string
                         }
                     }
 
-                    // We wait for <getAppIdTimeout> seconds (which is 0 at this point) to retrieve the appId. If retrieved during that time, we return success setting the correlation id.
-                    // If we are still waiting on the result beyond the timeout - for this particular call we return the failure but queue a task continuation for it to be cached for next time.
-                    Task<string> getAppIdTask = this.provideAppId(instrumentationKey.ToLowerInvariant());           
-                    if (getAppIdTask.Wait(GetAppIdTimeout))
-                    {
-                        this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, getAppIdTask.Result);
-                        correlationId = this.knownCorrelationIds[instrumentationKey];
-                        return true;
-                    }
-                    else
+                    // We only want one task to be there to fetch the ikey. If initial requests come in a bunch, only one of them gets to take responsibility of creating the fetch task. Rest return.
+                    if (this.FetchTasks.TryAdd(instrumentationKey, int.MinValue))
                     {
-                        getAppIdTask.ContinueWith((appId) =>
+                        try
                         {
-                            try
+                            // We wait for <getAppIdTimeout> seconds (which is 0 at this point) to retrieve the appId. If retrieved during that time, we return success setting the correlation id.
+                            // If we are still waiting on the result beyond the timeout - for this particular call we return the failure but queue a task continuation for it to be cached for next time.
+                            Task<string> getAppIdTask = this.provideAppId(instrumentationKey.ToLowerInvariant());
+                            this.FetchTasks[instrumentationKey] = getAppIdTask.Id;
+
+                            if (getAppIdTask.Wait(GetAppIdTimeout))
                             {
-                                this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, appId.Result);
+                                this.GenerateCorrelationIdAndAddToDictionary(instrumentationKey, getAppIdTask.Result);
+                                correlationId = this.knownCorrelationIds[instrumentationKey];
+                                return true;
                             }
-                            catch (Exception ex)
+                            else
                             {
-                                this.RegisterFailure(instrumentationKey, ex);
+                                getAppIdTask.ContinueWith((appId) =>
```

Comment:
That is totally fine. You can schedule a task with status=RanToCompletion and the continuation would run. I just did a quick test to double check as well.